rplidar.go

```diff
@@ -98,63 +102,88 @@ func newRplidar(ctx context.Context, _ resource.Dependencies, c resource.Config,
-	rp.started = true
+		deviceMutex:            &sync.Mutex{},
+		cacheMutex:             &sync.RWMutex{},
+		cacheBackgroundWorkers: &sync.WaitGroup{},
 
-	// S1 rplidars do not require the motor to be started before scanning can begin
-	if rplidarModelByteMap[rp.device.model] != "S1" {
-		rp.logger.Debug("starting motor")
-		rp.device.driver.StartMotor()
+		logger: logger,
+	}
+
+	// Setup RPLiDAR
+	if err := rp.setupRPLidar(ctx); err != nil {
+		return nil, errors.Wrap(err, "there was a problem setting up the rplidar")
 	}
-	rp.device.driver.StartScan(false, true)
-	rp.nodes = gen.New_measurementNodeHqArray(rp.nodeSize)
+
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	rp.cancelFunc = cancelFunc
+
+	// Start background caching of pointcloud data
+	rp.cacheBackgroundWorkers.Add(1)
+	go func() {
+		defer rp.cacheBackgroundWorkers.Done()
+		rp.cachePointCloudLoop(cancelCtx)
+	}()
 
 	return rp, nil
 }
 
-// NextPointCloud performs a scan on the rplidar and performs some filtering to clean up the data.
-func (rp *Rplidar) NextPointCloud(ctx context.Context) (pointcloud.PointCloud, error) {
-	rp.mu.Lock()
-	defer rp.mu.Unlock()
-
-	if rp.close {
-		return nil, errors.New("resource (rplidar) is closed")
+// setupRPLiDAR starts the motor, in necessary, and warms up the device, discard several scans to
+// ensure data returned to the user is valid.
+func (rp *Rplidar) setupRPLidar(ctx context.Context) error {
+	// Start the motor
+	// Note: S1 rplidars do not need to start the motor before scanning can begin
+	if rplidarModelByteMap[rp.device.model] != "S1" {
+		rp.logger.Debug("starting motor")
+		rp.device.driver.StartMotor()
 	}
 
-	if !rp.started {
-		return nil, errors.New("resource (rplidar) failed to initialize properly")
-	}
+	// Setup rplidar scan and scan once as per warmup procedure
+	rp.device.driver.StartScan(false, true)
+	rp.nodes = gen.New_measurementNodeHqArray(defaultNodeSize)
 
-	pc, err := rp.getPointCloud(ctx)
-	if err != nil {
-		return nil, err
+	goutils.SelectContextOrWait(ctx, time.Duration(defaultWarmupNumDiscardedScans)*time.Second)
+	if _, err := rp.scan(ctx, defaultWarmupNumDiscardedScans); err != nil {
+		return err
 	}
-	return pc, nil
+
+	return nil
 }
 
-// Images is a part of the camera interface but is not implemented for the rplidar.
-func (rp *Rplidar) Images(ctx context.Context) ([]camera.NamedImage, resource.ResponseMetadata, error) {
-	return nil, resource.ResponseMetadata{}, errors.New("images unimplemented")
+// cachePointCloudLoop is a background process that repeatedly gets point cloud data from the rplidar
+// and caches it for later access.
+func (rp *Rplidar) cachePointCloudLoop(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			pc, err := rp.scan(ctx, defaultNumScans)
+			if err != nil {
+				rp.logger.Debugf("issue getting pointcloud to cache: %v", err)
+			}
+
+			rp.cacheMutex.Lock()
+			rp.cachedPointCloud = pc
+			rp.cacheMutex.Unlock()
```

Comment:
I wanted to avoid delaying the returned of the cached pointcloud during the scan op. I can defer the Unlock() but would prefer to leave the scan out of it. Particularly with this looping so fast, we may be spending 90% of our time in that scan op that doesn't actively affect the cached value